File: examples/enterprise/remote_webhook.py
"""
Remote Webhook Handler
======================
Offload conversation logic to an external service. Instead of a local
Python function, pass a URL as ``on_message`` — Patter POSTs each
transcribed utterance to your backend and speaks the response.

This is ideal for microservice architectures where the LLM / business
logic runs on a separate server (e.g., behind a load balancer).

Requirements:
    pip install patter python-dotenv

Environment variables (.env):
    DEEPGRAM_API_KEY     - Deepgram key for speech-to-text
    ELEVENLABS_API_KEY   - ElevenLabs key for text-to-speech
    TWILIO_ACCOUNT_SID   - Twilio account SID
    TWILIO_AUTH_TOKEN    - Twilio auth token
    TWILIO_PHONE_NUMBER  - Your Twilio phone number (E.164 format)
    WEBHOOK_URL          - Public URL where Twilio can reach this server
"""

import asyncio
import os

from dotenv import load_dotenv

load_dotenv()

from patter import Patter

phone = Patter(
    mode="local",
    twilio_sid=os.getenv("TWILIO_ACCOUNT_SID"),
    twilio_token=os.getenv("TWILIO_AUTH_TOKEN"),
    phone_number=os.getenv("TWILIO_PHONE_NUMBER"),
    webhook_url=os.getenv("WEBHOOK_URL"),
)

agent = phone.agent(
    provider="pipeline",
    system_prompt="You are a helpful assistant.",  # Used as fallback context
    stt=Patter.deepgram(api_key=os.environ["DEEPGRAM_API_KEY"]),
    tts=Patter.elevenlabs(api_key=os.environ["ELEVENLABS_API_KEY"], voice="aria"),
    language="en",
)

# ── Remote webhook as on_message ─────────────────────────────────────
#
# When on_message is a URL string, Patter POSTs a JSON payload to it:
#
#   POST https://api.yourcompany.com/patter/message
#   Headers:
#     Content-Type: application/json
#     X-Patter-Signature: sha256=<HMAC-SHA256 of body using your webhook secret>
#
#   Body:
#     {
#       "text": "What are your business hours?",
#       "call_id": "CA1234567890abcdef",
#       "caller": "+15551234567",
#       "history": [
#         {"role": "user", "text": "Hi there"},
#         {"role": "assistant", "text": "Hello! How can I help?"},
#         {"role": "user", "text": "What are your business hours?"}
#       ]
#     }
#
#   Expected response: {"text": "We're open Monday to Friday, 9 AM to 5 PM."}
#
# Verify the X-Patter-Signature header to ensure the request came from
# Patter. Compute HMAC-SHA256 of the raw request body using your webhook
# secret and compare with the header value.
#
# For low-latency streaming, use a WebSocket URL instead:
#   on_message="wss://api.yourcompany.com/patter/stream"
#
# Patter sends the same JSON payload over the WebSocket and reads the
# response as a single text frame.

if __name__ == "__main__":
    print("Listening for calls (remote webhook mode)...")
    asyncio.run(
        phone.serve(
            agent,
            port=8000,
            on_message="https://api.yourcompany.com/patter/message",
        )
    )

File: examples/enterprise/remote_webhook.ts
/**
 * @file remote_webhook.ts
 * @description Route onMessage to a remote webhook or WebSocket endpoint.
 *
 * Instead of running LLM logic in-process, point onMessage at an external URL.
 * Patter POSTs user transcripts to your server and speaks whatever your
 * endpoint returns. This is ideal for:
 *   - Keeping LLM logic behind your own firewall
 *   - Sharing one message handler across multiple Patter instances
 *   - Environments where you cannot run custom code alongside Patter
 *
 * Webhook contract:
 *   POST https://api.yourcompany.com/patter/message
 *   Headers:  Content-Type: application/json
 *             X-Patter-Signature: <HMAC-SHA256 of body with your secret>
 *   Body:     { "transcript": "...", "callId": "...", "caller": "..." }
 *   Response: { "reply": "text to speak" }
 *
 * WebSocket alternative:
 *   Pass a "wss://..." URL instead. Patter opens a persistent socket and
 *   sends JSON frames with the same shape. Your server responds with
 *   { "reply": "..." } frames.
 *
 * Prerequisites:
 *   - TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, TWILIO_PHONE_NUMBER in .env
 *   - DEEPGRAM_API_KEY, ELEVENLABS_API_KEY in .env
 *   - WEBHOOK_URL pointing to this server (e.g. ngrok)
 *
 * Usage:
 *   npx ts-node remote_webhook.ts
 */

import { Patter, deepgram, elevenlabs } from "getpatter";
import dotenv from "dotenv";

dotenv.config();

const phone = new Patter({
  mode: "local",
  twilioSid: process.env.TWILIO_ACCOUNT_SID!,
  twilioToken: process.env.TWILIO_AUTH_TOKEN!,
  phoneNumber: process.env.TWILIO_PHONE_NUMBER!,
  webhookUrl: process.env.WEBHOOK_URL!,
});

const agent = phone.agent({
  systemPrompt: "You are an enterprise sales assistant.",
  provider: "pipeline",
  stt: deepgram({ apiKey: process.env.DEEPGRAM_API_KEY! }),
  tts: elevenlabs({ apiKey: process.env.ELEVENLABS_API_KEY!, voice: "rachel" }),
  firstMessage: "Hi there! I'd love to help you find the right plan.",
});

async function main(): Promise<void> {
  // Pass a URL string as onMessage — Patter handles HTTP POST / WebSocket
  // negotiation automatically. Your endpoint receives the transcript and
  // returns the reply text.
  await phone.serve({
    agent,
    port: 8000,
    onMessage: "https://api.yourcompany.com/patter/message",
    // WebSocket alternative:
    // onMessage: "wss://api.yourcompany.com/patter/ws",
    onCallStart: async (data) => {
      console.log(`Call started — ID: ${data.callId}`);
    },
    onCallEnd: async (data) => {
      console.log(`Call ended — duration: ${data.duration}s`);
    },
  });

  console.log("Remote webhook agent listening on port 8000");
}

main();
